### DISCUSSION TITLE
Observation regarding carrier frequency

### DISCUSSION THREAD

**Discussion by KingJimmy27** [bengtmartensson/Infrared4Arduino#63] [General]
For an NEC1 controller I am using an Arduino pro mini 3.3V @ 8MHz. 
I changed the static const frequency_t frequency = 38000U; in src/Nec1Renderer.h:
`
#include <IrSenderPwm.h>
#include <Nec1Renderer.h>

IrSender* irsend = IrSenderPwm::getInstance(true);

static void sendNec1(unsigned D, unsigned S, unsigned F, unsigned times) {
    const IrSignal* irSignal = Nec1Renderer::newIrSignal(D, S, F);
    irsend->sendIrSignal(*irSignal, times);
    delete irSignal;
}

void setup() {
}

void loop() 
{ 
    delay(1000);
    sendNec1(135U, 124U, 154U, 1);
}
`

I test the robustness of my code by running over and over for extended periods.
After a short period the programme starts sending random code and can onle be recovered by RST.
![image](https://user-images.githubusercontent.com/19304622/103368942-c9845000-4ac0-11eb-8643-c8eaabb77746.png)
If I reset the carrier to 38400 the prog runs ok with an occasional glitch but then starts re-sending the correct code.
I can avoid this by using a watchdog but there must be an error.

**Comment by bengtmartensson** [comment]
For all practical purposes, there is no difference between 38000 and 38400, as I wrote in the other thread.

Does the sketch below work?

```
#include <IrSenderPwm.h>
#include <Nec1Renderer.h>

IrSender* irsend = IrSenderPwm::getInstance(true);
const IrSignal* irSignal = Nec1Renderer::newIrSignal(135U, 124U, 154U);

static void sendNec1(unsigned times) {
irsend->sendIrSignal(*irSignal, times);
}

void setup() {
}

void loop()
{
delay(1000);
sendNec1(1);
}
```

**Comment by KingJimmy27** [comment]
Your sketch works at 38000kHz.
I also tested the RAW code generation type and that works at 38000kHz.
```
//#define Test1
//#define Test2
#define Test3

#ifdef Test1
#include <IrSenderPwm.h>
#include <Nec1Renderer.h>

IrSender* irsend = IrSenderPwm::getInstance(true);

static void sendNec1(unsigned D, unsigned S, unsigned F, unsigned times) {
    const IrSignal* irSignal = Nec1Renderer::newIrSignal(D, S, F);
    irsend->sendIrSignal(*irSignal, times);
    delete irSignal;
}

void setup() {}

void loop() 
{ 
    delay(1000);
    sendNec1(135U, 124U, 145U, 1);
}
#endif

#ifdef Test2
#include <IrSenderPwm.h>
#include <Nec1Renderer.h>

IrSender* irsend = IrSenderPwm::getInstance(true);
const IrSignal* irSignal = Nec1Renderer::newIrSignal(135U, 124U, 145U);

static void sendNec1(unsigned times) {
    irsend->sendIrSignal(*irSignal, times);
}

void setup() {}

void loop()
{
    delay(1000);
    sendNec1(1);
}
#endif

#ifdef Test3
#include <IrSenderPwm.h>
IrSender* irsend = IrSenderPwm::getInstance(true);

// Command #1: NEC1_D135S124F145
// Protocol: nec1, Parameters: S=124 D=135 F=145
const microseconds_t intro_NEC1_D135S124F145[] PROGMEM = { 9024U, 4512U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 1692U, 564U, 564U, 564U, 564U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 564U, 564U, 1692U, 564U, 564U, 564U, 564U, 564U, 564U, 564U, 1692U, 564U, 564U, 564U, 564U, 564U, 1692U, 564U, 564U, 564U, 1692U, 564U, 1692U, 564U, 1692U, 564U, 564U, 564U, 1692U, 564U, 1692U, 564U, 564U, 564U, 38628 };
const microseconds_t repeat_NEC1_D135S124F145[] PROGMEM = { 9024U, 2256U, 564U, 65535U };

static void sendRaw(const microseconds_t intro_P[], size_t lengthIntro, const microseconds_t repeat_P[], size_t lengthRepeat, frequency_t frequency, unsigned times) {
    microseconds_t intro[lengthIntro];
    microseconds_t repeat[lengthRepeat];
    memcpy_PF(intro, (uint_farptr_t)intro_P, sizeof(microseconds_t) * lengthIntro);
    memcpy_PF(repeat, (uint_farptr_t)repeat_P, sizeof(microseconds_t) * lengthRepeat);
    IrSignal irSignal(intro, lengthIntro, repeat, lengthRepeat, NULL, 0U, frequency);
    irsend->sendIrSignal(irSignal, times);
}

void setup() {}

void loop()
{
    delay(1000);
    sendRaw(intro_NEC1_D135S124F145, 68U, repeat_NEC1_D135S124F145, 4U, 38000U, 1);
}
#endif`
```

**Comment by bengtmartensson** [comment]
Please try the current state, just checked in a fix.

> Your sketch works at 38000kHz.

I assume you mean that you just checked with 38000, not that something is not working with 38400.

**Comment by KingJimmy27** [comment]
OK, all three of the sketches above now work at 38000Hz.
I have now finished my remote controller, thanks for all the advice and help.
Just one last (there always has to be one):
In my Test3 above the lines:
_microseconds_t intro[lengthIntro];
microseconds_t repeat[lengthRepeat];_
Gives me an error warning in my Visual Micro development environment:
_expresion must have a constant value.
The value of parameters "lengthintro" cannot be used as a constant._
I am not sure what that means but the programme works ignoring the error.

👍

**Comment by bengtmartensson** [comment]
Ok, sounds good. I will tag the current state and release it as a new version. Thank you very much for discovering the problems.

> ... Gives me an error warning in my Visual Micro development environment:
> expresion must have a constant value.
> The value of parameters "lengthintro" cannot be used as a constant.

Well, your code, not my code ;-). But I think that array sizes should be `const`s or `const_expr`s.

**Reply by bengtmartensson** [reply]
> I will tag the current state and release it as a new version.

Has been tagged as Version 1.1.1; available in the Arduino library manager.

**Comment by bengtmartensson** [comment]
~This is an issue with IrScrutinizer, not with the Infrared4Arduino project. Please instead open an issue at IrScrutinizer.~
